create a cloud server workflow

trevj opened the pull request:
This adds a button to the "connect with a friend" screen which helps you create a cloud server. When clicked, it:
 * prompts you to log into your DigitalOcean account
 * creates a server, using your DigitalOcean account
 * SSHes into the new server and runs the install script
 * shows the invitation URL for the new server

This is very much MVP. Several issues:
 * oftentimes, the new server cannot be SSHed into for several seconds (clicking the new button again will resume the install) https://github.com/uProxy/uproxy/issues/2190
 * the new server should be added to the roster
 * needs a much better success dialog (right now it's just a toast)
 * error handling throughout isn't great

For the dev build, only. @lucyhe: would appreciate your input here too.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2285)
<!-- Reviewable:end -->

dborkan commented:
Generally looking good!  Just a few small comments

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/generic_core/freedom-module.ts, line 79 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxugMIGomeNTlsgkcw:-KBxugMIGomeNTlsgkcx:937339751) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_core/freedom-module.ts#L79)):*
Any reason why you have it so that the ui calls core.deployCloudServer, then core.cloudInstall?  It seems to me that you would always want to do these things together, so if the ui just called something like core.deployAndInstallCloudServer we would only need 1 new command here and could clean up the CoreConnector, uproxy_core_api.ts, etc a little bit

---

*[src/generic_ui/polymer/cloud-install.ts, line 15 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBy1WTbBt-kzqa1Bul7:-KBy1WTbBt-kzqa1Bul8:1130038225) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/polymer/cloud-install.ts#L15)):*
can we call these something like openLoginDialog and closeAllDialogs?  Wizard is a cool name but it's not so obvious what it means

---

*[src/generic_ui/polymer/cloud-install.ts, line 31 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBy1vbAZp5l5JIeZ4RD:-KBy1vbAZp5l5JIeZ4RE:-982045164) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/polymer/cloud-install.ts#L31)):*
This is a UX question so we can deal with it later, but doesn't the toast quickly disappear after it's displayed?  We probably want to display the invite in some read-only input field, where we can automatically select the whole thing once the user clicks it.

---

*[src/generic_ui/scripts/ui.ts, line 317 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxwUfrkS1EpUN1TfXt:-KBxwUfs7UK48N68PTRQ:1939802181) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/scripts/ui.ts#L317)):*
Is there a risk that after you deploy a cloud server, this says something like "Install complete", then when you go to install a 2nd cloud server it hasn't been re-initialized and still says "Install complete"?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-:-KBy2KPoTVR5diaPWrKj:1511790193)*
<!-- Sent from Reviewable.io -->

trevj commented:
Many thanks! PTAL.

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/generic_core/freedom-module.ts, line 79 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxugMIGomeNTlsgkcw:-KC0yJvUk1Y9hocuqnfT:520257290) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_core/freedom-module.ts#L79)):*
I tried to explain this in `uproxy_core_api.ts` - basically, I wanted to separate the task of deploying a server from installing on a server because in future you could imagine the wizard accepting details for a box you *already have*, e.g. on a cloud computing provider we don't natively support.

That said, I can forsee a slippery slope of increasingly finer-grained method calls...and, like you say, they're not easy to add.

So, I've reduced the number of new calls to just one, `cloudInstall`, and added a return type. This should let us introduce finer-grained method calls in future, as we need them.

Thanks!

---

*[src/generic_ui/polymer/cloud-install.ts, line 15 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBy1WTbBt-kzqa1Bul7:-KC1CLlDhFVo3prfEQ5g:1321998317) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/polymer/cloud-install.ts#L15)):*
Well, it's a wizard in the sense that it guides the user through a sequence of steps - choosing the region will be one such step. But that would make more sense if we used something more wizard-like such as `core-pages`.

`openLoginDialog `and ` closeDialogs` sound good to me.

---

*[src/generic_ui/polymer/cloud-install.ts, line 31 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBy1vbAZp5l5JIeZ4RD:-KC1KhvfuJCTcyvOlwiv:1596756787) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/polymer/cloud-install.ts#L31)):*
Yes it does. I've replaced the toasts with a pair of dialogs. WDYT? The wording still needs some work.

---

*[src/generic_ui/scripts/ui.ts, line 317 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxwUfrkS1EpUN1TfXt:-KC1dw4IV1ORK-7RfQhq:-1564336424) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/scripts/ui.ts#L317)):*
Yes there is, at least for a microsecond. What would be a better way...trigger the appearance of `installingDialog` to changes in the `ui.cloudInstallStatus` field?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-:-KC1e78yzeaQf1j9ZZVI:-1794383077)*
<!-- Sent from Reviewable.io -->

dborkan commented:
One last comment, then :+1: 

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/generic_ui/scripts/ui.ts, line 317 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxwUfrkS1EpUN1TfXt:-KC2hjECbSLCFc3y3INa:-3074389) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/scripts/ui.ts#L317)):*
Yeah I think that would be fine to set ```ui.cloudInstallStatus = ''``` each time the ```installingDialog``` opens

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-:-KC2hsNPAjiVJVRMsMwz:-1675330965)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/generic_ui/scripts/ui.ts, line 317 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2285#-KBxwUfrkS1EpUN1TfXt:-KCGWFAMt6Y80znlMl68:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/2717d2d082ed323660391e563b643b0e6502577c/src/generic_ui/scripts/ui.ts#L317)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2285)*
<!-- Sent from Reviewable.io -->
